# Q1. 데이터를 로드한 뒤 상위 5개 컬럼을 출력해보자.

## 뉴욕 airBnB : https://www.kaggle.com/ptoscano230382/air-bnb-ny-2019 

In [93]:
import pandas as pd

air_df = pd.read_csv('./AB_NYC_2019.csv/AB_NYC_2019.csv')
air_df.head(5)

id                                              name  host_id  \
0  2539                Clean & quiet apt home by the park     2787   
1  2595                             Skylit Midtown Castle     2845   
2  3647               THE VILLAGE OF HARLEM....NEW YORK !     4632   
3  3831                   Cozy Entire Floor of Brownstone     4869   
4  5022  Entire Apt: Spacious Studio/Loft by central park     7192   

     host_name neighbourhood_group neighbourhood  latitude  longitude  \
0         John            Brooklyn    Kensington  40.64749  -73.97237   
1     Jennifer           Manhattan       Midtown  40.75362  -73.98377   
2    Elisabeth           Manhattan        Harlem  40.80902  -73.94190   
3  LisaRoxanne            Brooklyn  Clinton Hill  40.68514  -73.95976   
4        Laura           Manhattan   East Harlem  40.79851  -73.94399   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0     Private room    149               1                  9  2018-10-19   
1  Entire home/apt    225               1                 45  2019-05-21   
2     Private room    150               3                  0         NaN   
3  Entire home/apt     89               1                270  2019-07-05   
4  Entire home/apt     80              10                  9  2018-11-19   

   reviews_per_month  calculated_host_listings_count  availability_365  
0               0.21                               6               365  
1               0.38                               2               355  
2                NaN                               1               365  
3               4.64                               1               194  
4               0.10                               1                 0

# Q2. 각 host_name의 빈도수를 구하고 host_name으로 정렬하여 상위 5개를 출력해보자.

In [121]:
# 2차원
air_df.groupby(by='host_name').count().head(5)

# 1차원
air_df.groupby(by='host_name').size().head(5)

host_name
#NAME?                      1
'Cil                        1
(Ari) HENRY LEE             1
(Email hidden by Airbnb)    6
(Mary) Haiy                 1
dtype: int64

In [114]:
air_df.host_name.value_counts().sort_index().head(5)

#NAME?                      1
'Cil                        1
(Ari) HENRY LEE             1
(Email hidden by Airbnb)    6
(Mary) Haiy                 1
Name: host_name, dtype: int64

# Q3. 데이터의 각 host_name의 빈도수를 구하고, 
# 빈도수 기준 내림차순 정렬한 데이터 프레임을 생성해보자.
# 빈도수 컬럼의 컬럼명은 counts로 변경해보자.(host_name은 index입니당 ^^;)

In [141]:
# 1번째 방법 - value_counts()
air_freq_sr = air_df.host_name.value_counts()

# df 생성해보기
air_freq_df = pd.DataFrame(air_freq_sr)

# df의 컬럼명을 counts로 변경
air_freq_df.rename(columns={'host_name' : 'counts'}, inplace=True)
air_freq_df

counts
Michael               417
David                 403
Sonder (NYC)          327
John                  294
Alex                  279
...                   ...
Rhonycs                 1
Brandy-Courtney         1
Shanthony               1
Aurore And Jamila       1
Ilgar & Aysel           1

[11452 rows x 1 columns]

In [151]:
# 2번째 방법 - Groupby() :: index 명을 지우지 않는다..
air_df.groupby(by='host_name').size().to_frame().rename(columns={0 : 'counts'}).sort_values(by='counts', ascending=False)


counts
host_name           
Michael          417
David            403
Sonder (NYC)     327
John             294
Alex             279
...              ...
Jerbean            1
Jerald             1
Jeonghoon          1
Jeny               1
현선                 1

[11452 rows x 1 columns]

||counts|
|---|---|
|host_name||
|Michael|417|
|David|403|
|Sonder (NYC)|327|
|John|294|
|Alex|279|

# Q4. neighbourhood_group값에 따른 neighbourhood컬럼 값의 갯수를 구해보세요.

In [178]:
neighbourhood_df = pd.DataFrame(air_df.groupby(by=['neighbourhood_group', 'neighbourhood']).size())
neighbourhood_df.rename(columns={0 : 'size'})

size
neighbourhood_group neighbourhood      
Bronx               Allerton         42
                    Baychester        7
                    Belmont          24
                    Bronxdale        19
                    Castle Hill       9
...                                 ...
Staten Island       Tottenville       7
                    West Brighton    18
                    Westerleigh       2
                    Willowbrook       1
                    Woodrow           1

[221 rows x 1 columns]

|neighbourhood_group|neighbourhood|size|
|---|---|---|
|Bronx|Allerton|42|
|Bronx|Baychester|7|
|Bronx|Belmont|24|
|Bronx|Bronxdale|19|


# Q5. neighbourhood_group값에 따른 reviews_per_month 값의 평균, 분산, 최대, 최소값을 구해보자- agg함수 활용

In [184]:
reviews_df = air_df[['neighbourhood_group', 'reviews_per_month']].groupby(by='neighbourhood_group').agg(['mean', 'var', 'max', 'min'])
reviews_df

reviews_per_month                       
                                 mean       var    max   min
neighbourhood_group                                         
Bronx                        1.837831  2.799878  10.34  0.02
Brooklyn                     1.283212  2.299040  14.00  0.01
Manhattan                    1.272131  2.651206  58.50  0.01
Queens                       1.941200  4.897848  20.94  0.01
Staten Island                1.872580  2.840895  10.12  0.02

# Q6. neighbourhood_group값과 neighbourhood값에 따른 price의 평균을 구해보자.

In [193]:
air_df.groupby(['neighbourhood_group', 'neighbourhood']).price.mean()

neighbourhood_group  neighbourhood
Bronx                Allerton          87.595238
                     Baychester        75.428571
                     Belmont           77.125000
                     Bronxdale         57.105263
                     Castle Hill       63.000000
                                         ...    
Staten Island        Tottenville      144.857143
                     West Brighton     80.555556
                     Westerleigh       71.500000
                     Willowbrook      249.000000
                     Woodrow          700.000000
Name: price, Length: 221, dtype: float64

# 02. Apply_lambda
# 카드이용데이터 : https://www.kaggle.com/sakshigoyal7/credit-card-customers 

# Q1. 데이터를 로드하고, 데이터의 정보를 확인해보세요.

In [94]:
bank_df = pd.read_csv('./BankChurners.csv/BankChurners.csv')
bank_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10127 entries, 0 to 10126
Data columns (total 23 columns):
 #   Column                                                                                                                              Non-Null Count  Dtype  
---  ------                                                                                                                              --------------  -----  
 0   CLIENTNUM                                                                                                                           10127 non-null  int64  
 1   Attrition_Flag                                                                                                                      10127 non-null  object 
 2   Customer_Age                                                                                                                        10127 non-null  int64  
 3   Gender                                                                           

# Q2. Income_Category의 카테고리를 다음과 같이 변경하여 newIncome 컬럼에 넣어주세요 ^^ 
# map함수를 활용하셔도, apply를 활용하셔도 상관없습니다 ^^
```
Unknown : N
Less than $40K : a
$40K - $60K : b
$60K - $80K : c
$80K - $120K : d
$120K +’ : e
```

In [95]:
def names(name):
    if name == 'Unkown':
        return 'N'
    elif name == 'Less than $40K':
        return 'a'
    elif name == '$40K - $60K':
        return 'c'
    elif name == '$80K - $120K':
        return 'd'
    else:
        return 'e'
    
bank_df['newIncome'] = bank_df['Income_Category'].apply(lambda x : {x : names(x)})
bank_df

CLIENTNUM     Attrition_Flag  Customer_Age Gender  Dependent_count  \
0      768805383  Existing Customer            45      M                3   
1      818770008  Existing Customer            49      F                5   
2      713982108  Existing Customer            51      M                3   
3      769911858  Existing Customer            40      F                4   
4      709106358  Existing Customer            40      M                3   
...          ...                ...           ...    ...              ...   
10122  772366833  Existing Customer            50      M                2   
10123  710638233  Attrited Customer            41      M                2   
10124  716506083  Attrited Customer            44      F                1   
10125  717406983  Attrited Customer            30      M                2   
10126  714337233  Attrited Customer            43      F                2   

      Education_Level Marital_Status Income_Category Card_Category  \
0         High School        Married     $60K - $80K          Blue   
1            Graduate         Single  Less than $40K          Blue   
2            Graduate        Married    $80K - $120K          Blue   
3         High School        Unknown  Less than $40K          Blue   
4          Uneducated        Married     $60K - $80K          Blue   
...               ...            ...             ...           ...   
10122        Graduate         Single     $40K - $60K          Blue   
10123         Unknown       Divorced     $40K - $60K          Blue   
10124     High School        Married  Less than $40K          Blue   
10125        Graduate        Unknown     $40K - $60K          Blue   
10126        Graduate        Married  Less than $40K        Silver   

       Months_on_book  ...  Total_Revolving_Bal  Avg_Open_To_Buy  \
0                  39  ...                  777          11914.0   
1                  44  ...                  864           7392.0   
2                  36  ...                    0           3418.0   
3                  34  ...                 2517            796.0   
4                  21  ...                    0           4716.0   
...               ...  ...                  ...              ...   
10122              40  ...                 1851           2152.0   
10123              25  ...                 2186           2091.0   
10124              36  ...                    0           5409.0   
10125              36  ...                    0           5281.0   
10126              25  ...                 1961           8427.0   

       Total_Amt_Chng_Q4_Q1  Total_Trans_Amt  Total_Trans_Ct  \
0                     1.335             1144              42   
1                     1.541             1291              33   
2                     2.594             1887              20   
3                     1.405             1171              20   
4                     2.175              816              28   
...                     ...              ...             ...   
10122                 0.703            15476             117   
10123                 0.804             8764              69   
10124                 0.819            10291              60   
10125                 0.535             8395              62   
10126                 0.703            10294              61   

       Total_Ct_Chng_Q4_Q1  Avg_Utilization_Ratio  \
0                    1.625                  0.061   
1                    3.714                  0.105   
2                    2.333                  0.000   
3                    2.333                  0.760   
4                    2.500                  0.000   
...                    ...                    ...   
10122                0.857                  0.462   
10123                0.683                  0.511   
10124                0.818                  0.000   
10125                0.722                  0.000   
10126                0.649                  0.189   

       Naive_Bayes

# Q3. Customer_Age의 값을 이용하여 나이 구간을 Age_cat이라는 새 컬럼을 만들어주세요.
## (0~9:0 // 10~19:10 // 20~29:20...)

#  각 구간의 빈도수를 출력하라.

In [96]:
def custome_age(age):
    if age <= 9:
        return 0
    elif age <= 19:
        return 10
    elif age <= 29:
        return 20
    elif age <= 39:
        return 30
    elif age <= 49:
        return 40
    elif age <= 59:
        return 50
    elif age <= 69:
        return 60
    elif age <= 79:
        return 70

bank_df['Age_cat'] = bank_df['Customer_Age'].apply(lambda x : custome_age(x))
bank_df['Age_cat'].value_counts()

40    4561
50    2998
30    1841
60     530
20     195
70       2
Name: Age_cat, dtype: int64

# Q4. Education_Level의 값에서 Graduate이 포함된 경우에는 1 그렇지 않다면 0으로 변경하여
# new_edu_level 컬럼을 정의하고 빈도수를 출력해보세요.

In [100]:
bank_df['new_edu_level'] = bank_df['Education_Level'].str.contains('Graduate').apply(lambda x : 1 if x else 0)
bank_df['new_edu_level'].value_counts()

0    6483
1    3644
Name: new_edu_level, dtype: int64

# Q5. Marital_Status 컬럼값이 Married이고 Card_Category 컬럼의 값이 Platinum인 경우 1 그렇지 
# 않은 경우에는 모두 0으로 정의해주세요...

# 그리고 그 결과를 new_status 컬럼에 넣어주세요.

In [277]:
bank_df['newState'] = bank_df.apply(lambda x : 1 if x.Marital_Status == 'Married' and x.Card_Category == 'Platinum' else 0, axis=1)

bank_df['newState'].value_counts()


0    10120
1        7
Name: newState, dtype: int64

In [11]:
import numpy as np
import pandas as pd


datase = {
    '가' : np.arange(1, 5),
    '나' : np.arange(5, 9),
    '다' : np.arange(9, 13),
    '라' : np.arange(13, 17),
}

df = pd.DataFrame(datase)

# 0행의 데이터를 가져와 5보다 작다면 'N'으로 바꿔라
df['가'] = df.apply(lambda x : x if x.가 > 5 else 'N', axis=1)
df

가  나   다   라
0  N  5   9  13
1  N  6  10  14
2  N  7  11  15
3  N  8  12  16